### PR TITLE
fix(cli): Support PNPM when retrieving package dependency information

### LINKED ISSUE #2903: CLI: using pnpm causes multiple `npm list` errors written to CLI output for `provider`, `synth` and `debug` commands
### Expected Behavior

When the `cdktf-cli` package is installed as a dev dependency via `pnpm`, no `npm` errors are written to the output.

### Actual Behavior

When running some commands (`provider get`, `debug`, `synth`, multiple errors are written of the following form:

```log
[2023-05-31T23:22:01.376] [ERROR] default - npm
npm[2023-05-31T23:22:01.377] [ERROR] default -  ERR! code ELSPROBLEMS
npm ERR! invalid: constructs@10.2.24 /Users/[user]/src/[project]/ops/node_modules/constructs

 ERR! code ELSPROBLEMS
npm ERR! invalid: constructs@10.2.24 /Users/[user]/src/[project]/ops/node_modules/constructs
[2023-05-31T23:22:01.377] [ERROR] default -
npm ERR! A complete log of this run can be found in:
npm

npm ERR! A complete log of this run can be found in:
npm[2023-05-31T23:22:01.378] [ERROR] default -  ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_21_59_841Z-debug-0.log

 ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_21_59_841Z-debug-0.log
```

I'm uncertain if this is causing an issue or not: the command ends up succeeding.

### Steps to Reproduce

1. Create a typescript CDKTF project
2. Convert to the pnpm package manager (`pnpm import yarn.lock` or `pnpm import package-lock.json`)
3. `pnpm add cdktf-cli@latest`
4. Run `pnpm install`
5. Run `pnpm exec cdktf provider get` or `pnpm exec cdktf debug`

### Versions

```log
> [project]-cdk@0.1.0 cdktf /Users/[user]/src/[project]/ops
> cdktf "debug"

[2023-05-31T23:18:21.003] [ERROR] default - npm
npm[2023-05-31T23:18:21.004] [ERROR] default -  ERR! code ELSPROBLEMS
npm ERR! invalid: jsii@5.0.10 /Users/[user]/src/[project]/ops/node_modules/.pnpm/node_modules/jsii

 ERR! code ELSPROBLEMS
npm ERR! invalid: jsii@5.0.10 /Users/[user]/src/[project]/ops/node_modules/.pnpm/node_modules/jsii
[2023-05-31T23:18:21.005] [ERROR] default -


[2023-05-31T23:18:21.005] [ERROR] default - npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_18_19_473Z-debug-0.log

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_18_19_473Z-debug-0.log
[2023-05-31T23:18:21.006] [ERROR] default - npm
npm[2023-05-31T23:18:21.006] [ERROR] default -  ERR! code ELSPROBLEMS

 ERR! code ELSPROBLEMS
[2023-05-31T23:18:21.007] [ERROR] default - npm ERR!
npm ERR![2023-05-31T23:18:21.007] [ERROR] default -  invalid: constructs@10.2.24 /Users/[user]/src/[project]/ops/node_modules/constructs

 invalid: constructs@10.2.24 /Users/[user]/src/[project]/ops/node_modules/constructs
[2023-05-31T23:18:21.008] [ERROR] default -


[2023-05-31T23:18:21.008] [ERROR] default - npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_18_19_473Z-debug-0.log

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/[user]/.npm/_logs/2023-05-31T13_18_19_473Z-debug-0.log
cdktf debug
language: typescript
cdktf-cli: 0.16.1
node: v18.16.0
cdktf: 0.16.1
constructs: null
jsii: null
terraform: 1.4.6
arch: arm64
os: darwin 22.4.0
```

### Providers

```text
┌───────────────┬──────────────────┬─────────┬────────────┬─────────────────────┬─────────────────┐
│ Provider Name │ Provider Version │ CDKTF   │ Constraint │ Package Name        │ Package Version │
├───────────────┼──────────────────┼─────────┼────────────┼─────────────────────┼─────────────────┤
│ aws           │ 4.66.1           │ ^0.16.0 │            │ @cdktf/provider-aws │ 14.0.4          │
└───────────────┴──────────────────┴─────────┴────────────┴─────────────────────┴─────────────────┘
```

### Gist

https://gist.github.com/jamestelfer/03f844d1e1dc83df2c7d79730545a7a3

### Workarounds

I haven't seen this with a globally installed CLI, but I'm not 100% certain why.

We don't install the CLI globally as we want the local development environment consistent with CI, and we have multiple projects that don't necessarily change CLI versions at the same time.

### Anything Else?

The command being run (`npm list jsii --json` or similar) works with pnpm: `pnpm list jsii --json`. I haven't checked the JSON output.

If `pnpm` was run if a `pnpm-lock.yaml` is detected, this would avoid the error.

I don't know what the effect of this error is.

I'm willing to have a look at a fix, but I'm unsure of the timeline that I'd have to contribute it -- I might be able to get someone on my team to have a look.

### References

_No response_

### Help Wanted

- [X] I'm interested in contributing a fix myself

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by jamestelfer**
### Related issue

Fixes #2903 

### Description

`pnpm` uses a lock file called `pnpm-lock.yaml`, which can be used as a strong signal about which package manager is in use. There is an upcoming `packageManager` element in `package.json` that could also be used, but in practice this is not yet widespread.

Given the presence of this file, the updated logic switches to use `pnpm list` instead of `npm list`, and adapts to the updated file structure. I think it is fairly unlikely that a `pnpm-lock.yaml` file will exist if `pnpm` is not in use, so I've kept this very simple.

Further, I haven't made DRY efforts between the two methods, as I wasn't certain how useful that would be. I'm very happy to go further if the reviewers would like me to.

### Checklist

- [x] I have updated the PR title to match [CDKTF's style guide](https://github.com/hashicorp/terraform-cdk/blob/main/CONTRIBUTING.md#pull-requests-1)
- [x] I have run the linter on my code locally
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation if applicable
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works if applicable
- [x] New and existing unit tests pass locally with my changes

**Notes on tests:**

* no new unit tests are added -- there does not seem to be any tests around the existing functionality
* test cases run the same locally -- for me `main` is not clean, but the results between `main` and this branch are the same.
* I have tested projects that use yarn and pnpm with this change without issue

I'm happy to adjust the PR given feedback on any aspect.

### CODE REVIEW

**Comment by jamestelfer** [comment]
I have tested this locally against repositories that use yarn and pnpm as their package manager. The `debug` command returns consistent results in both cases now, there are no error messages emitted, and functionality (synth/deploy) etc works as expected.

The project using `pnpm` has no issues currently apart from these seemingly non-functional errors, and is happily in production.

It doesn't appear that there is explicit test coverage against this class, so I haven't added tests for the changes. Happy to take feedback on this.

**Comment by jamestelfer** [comment]
I have run tests locally against `main` and this branch, and they both appear to fail but in the same way. To the best of my knowledge this change doesn't regress current functionality.

**Comment by jamestelfer** [comment]
Note: I am looking to have the CLA signed on Monday, just double-checking everything with legal at work to make sure all the "T's" are crossed etc.

File: packages/@cdktf/commons/src/debug.ts [hunk 1]
```diff
@@ -72,8 +73,60 @@ export function getNodeVersion() {
...
+  }
+
+  if (
+    !json ||
+    !Array.isArray(json) ||
+    json.length === 0 ||
+    !json[0]?.dependencies?.[packageName]?.version
+  ) {
+    logger.debug(
+      `Unable to find '${packageName}' in 'npm list ${packageName} --json': ${output}`
```
Comment (ansgarm) [inline packages/@cdktf/commons/src/debug.ts:117]:
```suggestion
      `Unable to find '${packageName}' in 'pnpm list ${packageName} --json': ${output}`
```

**Review by ansgarm — COMMENTED** [review]
Thank you for the PR @jamestelfer! Found one small typo, but other than that it's good to go from my point of view 👍

**Review by DanielMSchmidt — APPROVED** [review]

**Comment by jamestelfer** [comment]
Thanks @ansgarm for fixing the typo and your review time! Really appreciate getting this small change in!

**Pull request merged**